src/utils/timing.rs

```diff
@@ -23,6 +25,62 @@ where
     result
 }
 
+// Measures the duration of an async operation and retries it if it fails.
+pub async fn measure_operation_with_retry<T, F, Fut>(
+    operation_name: &str,
+    f: &F,
+    timeout: Duration,
+    num_retries: u32,
+) -> anyhow::Result<T>
+where
+    F: Fn() -> Fut,
+    Fut: std::future::Future<Output = T>,
+{
+    trace!(operation = operation_name, "Starting operation");
+
+    for attempt in 0..=num_retries {
+        let start = Instant::now();
+
+        match time::timeout(timeout, f()).await {
```

Comment:
We might want to have some small back-off here, at least sleep for a second or two.